packages/backend-for-frontend/src/services/toolService.ts

```diff
@@ -0,0 +1,469 @@
+        failures: apiErrorsToValidationFailures(keyRetrieveErrors),
+      },
+      clientAssertionSignatureVerification: {
+        result: getStepResult(
+          [...clientAssertionErrors, ...keyRetrieveErrors],
+          clientSignatureErrors
+        ),
+        failures: apiErrorsToValidationFailures(clientSignatureErrors),
+      },
+      platformStatesVerification: {
+        result: getStepResult(
+          [
+            ...clientAssertionErrors,
+            ...keyRetrieveErrors,
+            ...clientSignatureErrors,
+          ],
+          platformStateErrors
+        ),
+        failures: apiErrorsToValidationFailures(platformStateErrors),
+      },
+    },
+  };
+}
+
+function assertIsConsumer(
+  requesterId: string,
+  keyWithClient: authorizationApi.KeyWithClient
+) {
+  if (requesterId !== keyWithClient.client.consumerId) {
+    throw organizationNotAllowed(keyWithClient.client.id);
+  }
+}
+
+function getStepResult(
+  prevStepErrors: Array<ApiError<string>>,
+  currentStepErrors: Array<ApiError<string>>
+): bffApi.TokenGenerationValidationStepResult {
+  if (currentStepErrors.length > 0) {
+    return bffApi.TokenGenerationValidationStepResult.Enum.FAILED;
+  } else if (prevStepErrors.length > 0) {
+    return bffApi.TokenGenerationValidationStepResult.Enum.SKIPPED;
+  } else {
+    return bffApi.TokenGenerationValidationStepResult.Enum.PASSED;
+  }
+}
+
+async function retrieveKey(
+  {
+    authorizationClient,
+    purposeProcessClient,
+    agreementProcessClient,
+    catalogProcessClient,
+  }: PagoPAInteropBeClients,
+  jwt: ClientAssertion,
+  ctx: WithLogger<BffAppContext>
+): Promise<
+  SuccessfulValidation<ApiKey | ConsumerKey> | FailedValidation<ErrorCodes>
+> {
+  if (!jwt.payload.purposeId) {
+    return {
+      data: undefined,
+      errors: [purposeIdNotFoundInClientAssertion()],
+    };
+  }
+  const keyWithClient = await authorizationClient.token
+    .getKeyWithClientByKeyId({
+      params: {
+        clientId: jwt.payload.sub,
+        keyId: jwt.header.kid,
+      },
+      headers: ctx.headers,
+    })
+    .catch(() => undefined);
+
+  if (!keyWithClient) {
+    return {
+      data: undefined,
+      errors: [
+        clientAssertionPublicKeyNotFound(jwt.header.kid, jwt.payload.sub),
+      ],
+    };
+  }
+
+  assertIsConsumer(ctx.authData.organizationId, keyWithClient);
+
+  const { encodedPem } = await authorizationClient.client.getClientKeyById({
+    headers: ctx.headers,
+    params: {
+      clientId: keyWithClient.client.id,
+      keyId: jwt.header.kid,
+    },
+  });
+
+  const purposeId = unsafeBrandId<PurposeId>(jwt.payload.purposeId);
+
+  if (keyWithClient.client.kind === authorizationApi.ClientKind.enum.API) {
+    return {
+      errors: undefined,
+      data: {
+        clientKind: authorizationApi.ClientKind.enum.API,
+        kid: jwt.header.kid,
+        algorithm: "RS256",
+        publicKey: encodedPem,
+        clientId: unsafeBrandId<ClientId>(jwt.payload.iss),
```

Comment:
```suggestion
        clientId: keyWithClient.client.id,
```
Note: this should have been `sub`